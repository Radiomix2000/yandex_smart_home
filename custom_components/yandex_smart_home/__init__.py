"""The Yandex Smart Home component."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.const import CONF_ID, CONF_PLATFORM, CONF_TOKEN, SERVICE_RELOAD
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.entityfilter import BASE_FILTER_SCHEMA, FILTER_SCHEMA, EntityFilter
from homeassistant.helpers.reload import async_integration_yaml_config
from homeassistant.helpers.service import async_register_admin_service
from homeassistant.helpers.typing import ConfigType
import voluptuous as vol

from . import config_validation as ycv, const
from .const import (
    CONF_DEVICES_DISCOVERED,
    CONF_LINKED_PLATFORMS,
    CONF_SKILL,
    CONF_USER_ID,
    DOMAIN,
    ConnectionType,
    EntityFilterSource,
)
from .entry_data import ConfigEntryData
from .helpers import SmartHomePlatform
from .http import async_register_http

if TYPE_CHECKING:
    from .cloud_stream import CloudStreamManager

_LOGGER = logging.getLogger(__name__)

ENTITY_PROPERTY_SCHEMA = vol.All(
    cv.has_at_least_one_key(
        const.CONF_ENTITY_PROPERTY_ENTITY,
        const.CONF_ENTITY_PROPERTY_ATTRIBUTE,
        const.CONF_ENTITY_PROPERTY_VALUE_TEMPLATE,
    ),
    vol.All(
        {
            vol.Required(const.CONF_ENTITY_PROPERTY_TYPE): vol.Schema(vol.All(str, ycv.property_type)),
            vol.Optional(const.CONF_ENTITY_PROPERTY_UNIT_OF_MEASUREMENT): cv.string,
            vol.Optional(const.CONF_ENTITY_PROPERTY_TARGET_UNIT_OF_MEASUREMENT): cv.string,
            vol.Optional(const.CONF_ENTITY_PROPERTY_ENTITY): cv.entity_id,
            vol.Optional(const.CONF_ENTITY_PROPERTY_ATTRIBUTE): cv.string,
            vol.Optional(const.CONF_ENTITY_PROPERTY_VALUE_TEMPLATE): cv.template,
        },
        ycv.property_attributes,
    ),
)


ENTITY_MODE_MAP_SCHEMA = vol.Schema(
    {vol.All(cv.string, ycv.mode_instance): vol.Schema({vol.All(cv.string, ycv.mode): [cv.string]})}
)

ENTITY_RANGE_SCHEMA = vol.Schema(
    {
        vol.Optional(const.CONF_ENTITY_RANGE_MAX): vol.All(vol.Coerce(float), vol.Range(min=-100.0, max=1000.0)),
        vol.Optional(const.CONF_ENTITY_RANGE_MIN): vol.All(vol.Coerce(float), vol.Range(min=-100.0, max=1000.0)),
        vol.Optional(const.CONF_ENTITY_RANGE_PRECISION): vol.All(vol.Coerce(float), vol.Range(min=-100.0, max=1000.0)),
    },
)

ENTITY_CUSTOM_MODE_SCHEMA = vol.Schema(
    {
        vol.All(cv.string, ycv.mode_instance): vol.Any(
            vol.All(
                {
                    vol.Optional(const.CONF_ENTITY_CUSTOM_MODE_SET_MODE): cv.SERVICE_SCHEMA,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_ENTITY_ID): cv.entity_id,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_ATTRIBUTE): cv.string,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_TEMPLATE): cv.template,
                },
                ycv.custom_capability_state,
            ),
            cv.boolean,
        )
    }
)

ENTITY_CUSTOM_RANGE_SCHEMA = vol.Schema(
    {
        vol.All(cv.string, ycv.range_instance): vol.Any(
            vol.All(
                {
                    vol.Optional(const.CONF_ENTITY_CUSTOM_RANGE_SET_VALUE): vol.Any(cv.SERVICE_SCHEMA),
                    vol.Optional(const.CONF_ENTITY_CUSTOM_RANGE_INCREASE_VALUE): vol.Any(cv.SERVICE_SCHEMA),
                    vol.Optional(const.CONF_ENTITY_CUSTOM_RANGE_DECREASE_VALUE): vol.Any(cv.SERVICE_SCHEMA),
                    vol.Optional(const.CONF_ENTITY_RANGE): ENTITY_RANGE_SCHEMA,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_ENTITY_ID): cv.entity_id,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_ATTRIBUTE): cv.string,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_TEMPLATE): cv.template,
                },
                ycv.custom_capability_state,
            ),
            cv.boolean,
        )
    }
)


ENTITY_CUSTOM_TOGGLE_SCHEMA = vol.Schema(
    {
        vol.All(cv.string, ycv.toggle_instance): vol.Any(
            vol.All(
                {
                    vol.Optional(const.CONF_ENTITY_CUSTOM_TOGGLE_TURN_ON): cv.SERVICE_SCHEMA,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_TOGGLE_TURN_OFF): cv.SERVICE_SCHEMA,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_ENTITY_ID): cv.entity_id,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_ATTRIBUTE): cv.string,
                    vol.Optional(const.CONF_ENTITY_CUSTOM_CAPABILITY_STATE_TEMPLATE): cv.template,
                },
                ycv.custom_capability_state,
            ),
            cv.boolean,
        )
    }
)


ENTITY_SCHEMA = vol.All(
    vol.Schema(
        {
            vol.Optional(const.CONF_NAME): cv.string,
            vol.Optional(const.CONF_ROOM): cv.string,
            vol.Optional(const.CONF_TYPE): vol.All(cv.string, ycv.device_type),
            vol.Optional(const.CONF_TURN_ON): vol.Any(cv.SERVICE_SCHEMA, cv.boolean),
            vol.Optional(const.CONF_TURN_OFF): vol.Any(cv.SERVICE_SCHEMA, cv.boolean),
            vol.Optional(const.CONF_DEVICE_CLASS): vol.In(const.DEVICE_CLASS_BUTTON),
            vol.Optional(const.CONF_FEATURES): vol.All(cv.ensure_list, ycv.entity_features),
            vol.Optional(const.CONF_ENTITY_PROPERTIES): [ENTITY_PROPERTY_SCHEMA],
            vol.Optional(const.CONF_SUPPORT_SET_CHANNEL): cv.boolean,
            vol.Optional(const.CONF_STATE_UNKNOWN): cv.boolean,
            vol.Optional(const.CONF_COLOR_PROFILE): cv.string,
            vol.Optional(const.CONF_ERROR_CODE_TEMPLATE): cv.template,
            vol.Optional(const.CONF_ENTITY_RANGE): ENTITY_RANGE_SCHEMA,
            vol.Optional(const.CONF_ENTITY_MODE_MAP): ENTITY_MODE_MAP_SCHEMA,
            vol.Optional(const.CONF_ENTITY_CUSTOM_MODES): ENTITY_CUSTOM_MODE_SCHEMA,
            vol.Optional(const.CONF_ENTITY_CUSTOM_TOGGLES): ENTITY_CUSTOM_TOGGLE_SCHEMA,
            vol.Optional(const.CONF_ENTITY_CUSTOM_RANGES): ENTITY_CUSTOM_RANGE_SCHEMA,
        }
    )
)

NOTIFIER_SCHEMA = vol.Schema(
    {
        vol.Required(const.CONF_NOTIFIER_OAUTH_TOKEN): cv.string,
        vol.Required(const.CONF_NOTIFIER_SKILL_ID): cv.string,
        vol.Required(const.CONF_NOTIFIER_USER_ID): cv.string,
    },
)


SETTINGS_SCHEMA = vol.All(
    cv.deprecated(const.CONF_PRESSURE_UNIT),
    {
        vol.Optional(const.CONF_PRESSURE_UNIT): cv.string,
        vol.Optional(const.CONF_BETA): cv.boolean,
        vol.Optional(const.CONF_CLOUD_STREAM): cv.boolean,
    },
)


YANDEX_SMART_HOME_SCHEMA = vol.All(
    vol.Schema(
        {
            vol.Optional(const.CONF_NOTIFIER): vol.All(cv.ensure_list, [NOTIFIER_SCHEMA]),
            vol.Optional(const.CONF_SETTINGS): vol.All(lambda value: value or {}, SETTINGS_SCHEMA),
            vol.Optional(const.CONF_FILTER): BASE_FILTER_SCHEMA,
            vol.Optional(const.CONF_ENTITY_CONFIG): vol.All(lambda value: value or {}, {cv.entity_id: ENTITY_SCHEMA}),
            vol.Optional(const.CONF_COLOR_PROFILE): vol.Schema(
                {cv.string: {vol.All(ycv.color_name): vol.All(ycv.color_value)}}
            ),
        },
    )
)

CONFIG_SCHEMA = vol.Schema({DOMAIN: YANDEX_SMART_HOME_SCHEMA}, extra=vol.ALLOW_EXTRA)


class YandexSmartHome:
    """Yandex Smart Home component main class."""

    def __init__(self, hass: HomeAssistant, yaml_config: ConfigType):
        """Initialize the Yandex Smart Home from yaml configuration."""
        self.cloud_streams: dict[str, CloudStreamManager] = {}

        self._hass = hass
        self._yaml_config = yaml_config
        self._entry_datas: dict[str, ConfigEntryData] = {}

        async_register_admin_service(hass, DOMAIN, SERVICE_RELOAD, self._handle_yaml_config_reload)

    async def _handle_yaml_config_reload(self, _: Any) -> None:
        """Handle yaml configuration reloading."""
        if config := await async_integration_yaml_config(self._hass, DOMAIN):
            self._yaml_config = config.get(DOMAIN, {})

        for entry in self._hass.config_entries.async_entries(DOMAIN):
            await _async_entry_update_listener(self._hass, entry)

        return None

    def get_entry_data(self, entry: ConfigEntry) -> ConfigEntryData:
        """Return a config entry data for a config entry."""
        return self._entry_datas[entry.entry_id]

    def get_direct_connection_entry_data(
        self, platform: SmartHomePlatform, user_id: str | None
    ) -> ConfigEntryData | None:
        """Return a config entry data with direct connection config entry."""
        for data in self._entry_datas.values():
            if (
                data.connection_type == ConnectionType.DIRECT
                and data.entry.state == ConfigEntryState.LOADED
                and data.platform == platform
            ):
                if user_id and data.skill and data.skill.user_id == user_id:
                    return data
                if not user_id:
                    return data

        return None

    def get_diagnostics(self) -> ConfigType:
        """Return diagnostics for the component."""
        from homeassistant.components.diagnostics import async_redact_data

        return {"yaml_config": async_redact_data(self._yaml_config, [const.CONF_NOTIFIER])}

    async def async_setup_entry(self, entry: ConfigEntry) -> bool:
        """Set up a config entry."""
        entity_config = self._yaml_config.get(const.CONF_ENTITY_CONFIG)

        entity_filter: EntityFilter | None = None
        if entry.options.get(const.CONF_FILTER_SOURCE) == EntityFilterSource.YAML:
            if entity_filter_config := self._yaml_config.get(const.CONF_FILTER):
                entity_filter = FILTER_SCHEMA(entity_filter_config)
        else:
            entity_filter = FILTER_SCHEMA(entry.options.get(const.CONF_FILTER, {}))

        data = ConfigEntryData(
            hass=self._hass,
            entry=entry,
            yaml_config=self._yaml_config,
            entity_config=entity_config,
            entity_filter=entity_filter,
        )

        self._entry_datas[entry.entry_id] = await data.async_setup()
        entry.async_on_unload(entry.add_update_listener(_async_entry_update_listener))

        return True

    async def async_unload_entry(self, entry: ConfigEntry) -> bool:
        """Unload a config entry."""
        data = self.get_entry_data(entry)
        await data.async_unload()
        return True

    async def async_remove_entry(self, entry: ConfigEntry) -> None:
        """Remove a config entry."""
        try:
            del self._entry_datas[entry.entry_id]
        except KeyError:
            pass

        return None


async def async_setup(hass: HomeAssistant, yaml_config: ConfigType) -> bool:
    """Activate Yandex Smart Home component."""
    hass.data[DOMAIN] = component = YandexSmartHome(hass, yaml_config.get(DOMAIN, {}))
    async_register_http(hass, component)
    return True


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up a config entry."""
    component: YandexSmartHome = hass.data[DOMAIN]
    return await component.async_setup_entry(entry)


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    component: YandexSmartHome = hass.data[DOMAIN]
    return await component.async_unload_entry(entry)


async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Migrate the config entry upon new versions."""
    version = entry.version
    component: YandexSmartHome = hass.data[DOMAIN]
    data: ConfigType = {**entry.data}
    options: ConfigType = {**entry.options}

    _LOGGER.debug(f"Migrating from version {version}")

    if version == 1:
        preserve_keys = [
            const.CONF_CONNECTION_TYPE,
            const.CONF_CLOUD_INSTANCE,
            const.CONF_DEVICES_DISCOVERED,
            const.CONF_FILTER,
            const.CONF_USER_ID,
        ]
        for store in [data, options]:
            for key in list(store.keys()):
                if key not in preserve_keys:
                    store.pop(key, None)

        data.setdefault(const.CONF_CONNECTION_TYPE, ConnectionType.DIRECT)
        data.setdefault(const.CONF_DEVICES_DISCOVERED, True)

        version = 2
        hass.config_entries.async_update_entry(entry, data=data, options=options, version=version)
        _LOGGER.debug(f"Migration to version {version} successful")

    if version == 2:
        version = 3
        _LOGGER.debug(f"Migration to version {version} successful")

    if version == 3:
        options[const.CONF_FILTER_SOURCE] = EntityFilterSource.CONFIG_ENTRY
        if const.CONF_FILTER in component._yaml_config:
            options[const.CONF_FILTER_SOURCE] = EntityFilterSource.YAML

        version = 4
        hass.config_entries.async_update_entry(entry, data=data, options=options, version=version)
        _LOGGER.debug(f"Migration to version {version} successful")

    if version == 4:
        from .config_flow import DEFAULT_CONFIG_ENTRY_TITLE, PRE_V1_DIRECT_CONFIG_ENTRY_TITLE, async_config_entry_title

        title = entry.title
        data.setdefault(CONF_PLATFORM, SmartHomePlatform.YANDEX)

        if (
            len(hass.config_entries.async_entries(DOMAIN)) == 1
            and data[const.CONF_CONNECTION_TYPE] == ConnectionType.DIRECT
        ):
            for notifier_config in component._yaml_config.get(const.CONF_NOTIFIER, []):
                options.setdefault(
                    CONF_SKILL,
                    {
                        CONF_USER_ID: notifier_config[const.CONF_NOTIFIER_USER_ID],
                        CONF_ID: notifier_config[const.CONF_NOTIFIER_SKILL_ID],
                        CONF_TOKEN: notifier_config[const.CONF_NOTIFIER_OAUTH_TOKEN],
                    },
                )
                break

        if entry.title in (DEFAULT_CONFIG_ENTRY_TITLE, PRE_V1_DIRECT_CONFIG_ENTRY_TITLE):
            title = await async_config_entry_title(hass, data, options)

        version = 5
        hass.config_entries.async_update_entry(
            entry,
            title=title,
            data=data,
            options=options,
            version=version,
        )
        _LOGGER.debug(f"Migration to version {version} successful")

    if version == 5:
        if data.get(CONF_DEVICES_DISCOVERED):
            data[CONF_LINKED_PLATFORMS] = [SmartHomePlatform.YANDEX]

        version = 6
        hass.config_entries.async_update_entry(entry, data=data, version=version)
        _LOGGER.debug(f"Migration to version {version} successful")

    return True


async def async_remove_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Remove a config entry."""
    component: YandexSmartHome | None = hass.data.get(DOMAIN)
    if component:
        await component.async_remove_entry(entry)

    return None


async def _async_entry_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Handle config entry options update."""
    await hass.config_entries.async_reload(entry.entry_id)
    return None
