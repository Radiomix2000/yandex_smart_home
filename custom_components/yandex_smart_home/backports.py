"""Backports from newer Home Assistant versions."""

from enum import StrEnum


class LockState(StrEnum):
    """State of lock entities (2024.10+)."""

    JAMMED = "jammed"
    OPENING = "opening"
    LOCKING = "locking"
    OPEN = "open"
    UNLOCKING = "unlocking"
    LOCKED = "locked"
    UNLOCKED = "unlocked"
