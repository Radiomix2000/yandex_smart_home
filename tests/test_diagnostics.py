from datetime import datetime
from http import HTTPStatus
from unittest import mock

from homeassistant.const import CONF_ID, CONF_PLATFORM, CONF_TOKEN
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entityfilter, issue_registry
from homeassistant.helpers.issue_registry import IssueSeverity
from homeassistant.setup import async_setup_component
from pytest_homeassistant_custom_component.common import MockConfigEntry, MockUser
from pytest_homeassistant_custom_component.typing import ClientSessionGenerator
from syrupy import SnapshotAssertion

from custom_components.yandex_smart_home import DOMAIN
from custom_components.yandex_smart_home.config_flow import ConfigFlowHandler
from custom_components.yandex_smart_home.const import (
    CONF_CLOUD_INSTANCE,
    CONF_CLOUD_INSTANCE_PASSWORD,
    CONF_CONNECTION_TYPE,
    CONF_FILTER,
    CONF_FILTER_SOURCE,
    CONF_SKILL,
    CONF_USER_ID,
    ConnectionType,
    EntityFilterSource,
)
from custom_components.yandex_smart_home.helpers import SmartHomePlatform


async def test_diagnostics(
    hass_platform: HomeAssistant,
    hass_client: ClientSessionGenerator,
    hass_admin_user: MockUser,
    snapshot: SnapshotAssertion,
) -> None:
    hass = hass_platform
    yaml_config = {
        "filter": {"include_domains": ["light", "sensor", "binary_sensor"]},
        "entity_config": {
            "switch.with_template": {"error_code_template": '{{ "a" + "b" }}'},
            "light.kitchen": {"properties": [{"type": "temperature", "entity": "sensor.invalid"}]},
        },
        "notifier": [{"skill_id": "foo", "oauth_token": "token", "user_id": hass_admin_user.id}],
    }
    assert await async_setup_component(hass, "diagnostics", {})
    assert await async_setup_component(hass, DOMAIN, {DOMAIN: yaml_config})

    hass.states.async_set("sensor.invalid", "foo")
    config_entry = MockConfigEntry(
        entry_id="fe76008998bdad631c33d60ef044b9ac",
        domain=DOMAIN,
        version=ConfigFlowHandler.VERSION,
        data={
            CONF_CONNECTION_TYPE: ConnectionType.DIRECT,
            CONF_CLOUD_INSTANCE: {CONF_CLOUD_INSTANCE_PASSWORD: "foo"},
            CONF_PLATFORM: SmartHomePlatform.YANDEX,
        },
        options={
            CONF_FILTER_SOURCE: EntityFilterSource.CONFIG_ENTRY,
            CONF_FILTER: {entityfilter.CONF_INCLUDE_ENTITY_GLOBS: ["*"]},
            CONF_SKILL: {CONF_ID: "skill_id", CONF_TOKEN: "oauth_token", CONF_USER_ID: "user_id"},
        },
    )
    config_entry.add_to_hass(hass)

    now = datetime(2024, 5, 7, 1, 10, 6)
    with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
        await hass.config_entries.async_setup(config_entry.entry_id)
        await hass.async_block_till_done()

        issue_registry.async_create_issue(
            hass, DOMAIN, "foo", is_fixable=False, severity=IssueSeverity.CRITICAL, translation_key="foo"
        )

    client = await hass_client()
    response = await client.get(f"/api/diagnostics/config_entry/{config_entry.entry_id}")
    assert response.status == HTTPStatus.OK
    diagnostics = await response.json()
    for k in ("integration_manifest", "custom_components", "home_assistant", "minor_version", "setup_times"):
        diagnostics.pop(k, None)

    for k in ("minor_version", "created_at", "discovery_keys", "modified_at"):
        diagnostics["data"]["entry"].pop(k, None)

    assert diagnostics == snapshot
