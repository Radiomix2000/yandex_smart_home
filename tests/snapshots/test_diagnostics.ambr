# serializer version: 1
# name: test_diagnostics
  dict({
    'data': dict({
      'devices': dict({
        'binary_sensor.front_door': dict({
          'capabilities': list([
          ]),
          'description': dict({
            'device_info': dict({
              'model': 'binary_sensor.front_door',
            }),
            'id': 'binary_sensor.front_door',
            'name': 'Front Door',
            'properties': list([
              dict({
                'parameters': dict({
                  'events': list([
                    dict({
                      'value': 'opened',
                    }),
                    dict({
                      'value': 'closed',
                    }),
                  ]),
                  'instance': 'open',
                }),
                'reportable': True,
                'retrievable': True,
                'type': 'devices.properties.event',
              }),
            ]),
            'type': 'devices.types.sensor.open',
          }),
          'properties': list([
            '<OpenStateEventProperty device_id=binary_sensor.front_door type=devices.properties.event instance=open>',
          ]),
          'state': dict({
            'id': 'binary_sensor.front_door',
            'properties': list([
              dict({
                'state': dict({
                  'instance': 'open',
                  'value': 'opened',
                }),
                'type': 'devices.properties.event',
              }),
            ]),
          }),
        }),
        'light.kitchen': dict({
          'capabilities': list([
            '<ColorSettingCapability device_id=light.kitchen type=devices.capabilities.color_setting instance=base>',
            '<RGBColorCapability device_id=light.kitchen type=devices.capabilities.color_setting instance=rgb>',
            '<ColorTemperatureCapability device_id=light.kitchen type=devices.capabilities.color_setting instance=temperature_k>',
            '<OnOffCapabilityBasic device_id=light.kitchen type=devices.capabilities.on_off instance=on>',
            '<BrightnessCapability device_id=light.kitchen type=devices.capabilities.range instance=brightness>',
          ]),
          'description': dict({
            'capabilities': list([
              dict({
                'parameters': dict({
                  'color_model': 'rgb',
                  'temperature_k': dict({
                    'max': 6500,
                    'min': 1500,
                  }),
                }),
                'reportable': True,
                'retrievable': True,
                'type': 'devices.capabilities.color_setting',
              }),
              dict({
                'reportable': True,
                'retrievable': True,
                'type': 'devices.capabilities.on_off',
              }),
              dict({
                'parameters': dict({
                  'instance': 'brightness',
                  'random_access': True,
                  'range': dict({
                    'max': 100.0,
                    'min': 1.0,
                    'precision': 1.0,
                  }),
                  'unit': 'unit.percent',
                }),
                'reportable': True,
                'retrievable': True,
                'type': 'devices.capabilities.range',
              }),
            ]),
            'device_info': dict({
              'model': 'light.kitchen',
            }),
            'id': 'light.kitchen',
            'name': 'Kitchen Light',
            'properties': list([
              dict({
                'parameters': dict({
                  'instance': 'temperature',
                  'unit': 'unit.temperature.celsius',
                }),
                'reportable': True,
                'retrievable': True,
                'type': 'devices.properties.float',
              }),
            ]),
            'type': 'devices.types.light',
          }),
          'properties': list([
            "<TemperatureCustomFloatProperty device_id=light.kitchen instance=temperature value_template=Template<template=({{ states('sensor.invalid') }}) renders=2> value=UndefinedType._singleton>",
          ]),
          'state': dict({
            'capabilities': list([
              dict({
                'state': dict({
                  'instance': 'temperature_k',
                  'value': 4200,
                }),
                'type': 'devices.capabilities.color_setting',
              }),
              dict({
                'state': dict({
                  'instance': 'on',
                  'value': True,
                }),
                'type': 'devices.capabilities.on_off',
              }),
              dict({
                'state': dict({
                  'instance': 'brightness',
                  'value': 70,
                }),
                'type': 'devices.capabilities.range',
              }),
            ]),
            'id': 'light.kitchen',
          }),
        }),
        'sensor.invalid': dict({
          'capabilities': list([
          ]),
          'description': None,
          'properties': list([
          ]),
          'state': dict({
            'error_code': 'DEVICE_UNREACHABLE',
            'id': 'sensor.invalid',
          }),
        }),
        'sensor.outside_temp': dict({
          'capabilities': list([
          ]),
          'description': dict({
            'device_info': dict({
              'model': 'sensor.outside_temp',
            }),
            'id': 'sensor.outside_temp',
            'name': 'Температура за бортом',
            'properties': list([
              dict({
                'parameters': dict({
                  'instance': 'temperature',
                  'unit': 'unit.temperature.celsius',
                }),
                'reportable': True,
                'retrievable': True,
                'type': 'devices.properties.float',
              }),
            ]),
            'type': 'devices.types.sensor.climate',
          }),
          'properties': list([
            '<TemperatureSensor device_id=sensor.outside_temp type=devices.properties.float instance=temperature>',
          ]),
          'state': dict({
            'id': 'sensor.outside_temp',
            'properties': list([
              dict({
                'state': dict({
                  'instance': 'temperature',
                  'value': 15.6,
                }),
                'type': 'devices.properties.float',
              }),
            ]),
          }),
        }),
      }),
      'entry': dict({
        'data': dict({
          'cloud_instance': '**REDACTED**',
          'connection_type': 'direct',
          'platform': 'yandex',
        }),
        'disabled_by': None,
        'domain': 'yandex_smart_home',
        'entry_id': 'fe76008998bdad631c33d60ef044b9ac',
        'options': dict({
          'filter': dict({
            'include_entity_globs': list([
              '*',
            ]),
          }),
          'filter_source': 'config_entry',
          'skill': '**REDACTED**',
        }),
        'pref_disable_new_entities': False,
        'pref_disable_polling': False,
        'source': 'user',
        'title': 'Mock Title',
        'unique_id': None,
        'version': 6,
      }),
      'issues': list([
        dict({
          'created': '2024-05-07T01:10:06',
          'dismissed_version': None,
          'domain': 'yandex_smart_home',
          'is_persistent': False,
          'issue_id': 'deprecated_yaml_notifier',
        }),
        dict({
          'created': '2024-05-07T01:10:06',
          'dismissed_version': None,
          'domain': 'yandex_smart_home',
          'is_persistent': False,
          'issue_id': 'foo',
        }),
      ]),
      'yaml_config': dict({
        'entity_config': dict({
          'light.kitchen': dict({
            'properties': list([
              dict({
                'entity': 'sensor.invalid',
                'type': 'temperature',
              }),
            ]),
          }),
          'switch.with_template': dict({
            'error_code_template': dict({
              '__type': "<class 'homeassistant.helpers.template.Template'>",
              'repr': 'Template<template=({{ "a" + "b" }}) renders=0>',
            }),
          }),
        }),
        'filter': dict({
          'exclude_domains': list([
          ]),
          'exclude_entities': list([
          ]),
          'exclude_entity_globs': list([
          ]),
          'include_domains': list([
            'light',
            'sensor',
            'binary_sensor',
          ]),
          'include_entities': list([
          ]),
          'include_entity_globs': list([
          ]),
        }),
        'notifier': '**REDACTED**',
      }),
    }),
  })
# ---
